modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/AbstractExternalResult.java

```diff
@@ -0,0 +1,392 @@
+     */
+    @NotNull protected ByteBuffer readRowHeaderFromFile() {
+        return readDataFromFile(ROW_HEADER_SIZE);
+    }
+
+    /**
+     * Reads data from file.
+     *
+     * @param rowLen Data length to read.
+     * @return Byte buffer with data.
+     */
+    @NotNull protected ByteBuffer readDataFromFile(int rowLen) {
+        try {
+            ByteBuffer rowBytes = ByteBuffer.allocate(rowLen); // 2 x integer for length in bytes and columns count.
+
+            while (rowBytes.hasRemaining()) {
+                int bytesRead = fileIo.read(rowBytes);
+
+                if (bytesRead <= 0)
+                    throw new IOException("Can not read data from file: " + file.getAbsolutePath() +
+                        ", curPos=" + fileIo.position() + ", rowLen=" + rowLen + "]");
+            }
+
+            rowBytes.flip();
+
+            return rowBytes;
+        }
+        catch (IOException e) {
+            close();
+
+            throw new IgniteException("Failed to read query result the from spill file.", e);
+        }
+    }
+
+    /**
+     * @param cap Buffer initial capacity.
+     * @return New data buffer.
+     */
+    @NotNull protected Data createDataBuffer(int cap) {
+        return Data.create(null, cap, true);
+    }
+
+    /**
+     * Adds row to buffer.
+     *
+     * @param row Row.
+     * @param buff Buffer.
+     */
+    protected void addRowToBuffer(Value[] row, Data buff) {
+        int initPos = buff.length();
+
+        buff.checkCapacity(rowSize(row));
+
+        buff.writeInt(0); // Skip int position for row length in bytes.
+        buff.writeInt(row.length); // Skip int position for columns count.
+
+        for (int i = 0; i < row.length; i++)
+            buff.writeValue(row[i]);
+
+        int len = buff.length() - initPos - ROW_HEADER_SIZE;
+
+        buff.setInt(initPos, len);
+    }
+
+    /**
+     * Writes buffer to file.
+     *
+     * @param buff Buffer.
+     * @return Bytes written.
+     */
+    protected int writeBufferToFile(Data buff) {
+        try {
+            ByteBuffer byteBuff = ByteBuffer.wrap(buff.getBytes());
+
+            byteBuff.limit(buff.length());
+
+            while (byteBuff.hasRemaining()) {
+                int bytesWritten = fileIo.write(byteBuff);
+
+                if (bytesWritten <= 0)
+                    throw new IOException("Can not write data to file: " + file.getAbsolutePath());
+            }
+
+            return byteBuff.limit();
+        }
+        catch (IOException e) {
+            close();
+
+            throw new IgniteException("Failed to write intermediate query result to the spill file.", e);
+        }
+    }
+
+    /**
+     * Marks row in the file as removed by setting {@code -1} on it's {@code colCount} position.
+     *
+     * @param addr Row absolute address in the file.
+     */
+    protected void markRowRemoved(long addr) {
+        setFilePosition(addr);
+
+        ByteBuffer hdr = readRowHeaderFromFile();
```

Comment:
Could you please point out where I  read row from disk? I read only header.

Reply:
It doesn't matter what you read. My question is why do you read at least something from disk? As I can see you don't use this loaded data at all. It would make sense if you are doing the consistency checking of data but you don't have such a feature.